### PR TITLE
webui: Add Firefox theme for use on Live media

### PR DESCRIPTION
**Pull request by M4rtinK**
The theme upstream is here: https://github.com/garrett/firefox-theme-frame

Its main aim is to shape the Firefox window to the most suitable state on Live images:
- has a header bar for window switching
- has no tab/status/URL bars
- has no browser controls

As Firefox needs the profile folder to be writable, we need to copy the profile template to a writable tempdir & use that as the profile path for Firefox.

To simplify things, we also use the same theme on non Live images for now, with the expectation that a tweaked theme will provided later.

**TODO**
- [x] check the license if fine (https://github.com/garrett/firefox-theme-frame/blob/main/LICENSE) - is this compatible with our GPLv2 + MIT ?
- [x] drop the status bubble
- [x] drop the tab bar
- [x] fix the content not loading

### CODE REVIEW

**Comment by M4rtinK** [comment]
So far I've noticed a couple issues with this that will have to be resolved.

On Live images built via the Image builder pipeline, the Web UI shows up ~like we want it:
live_image_custom_theme_1

There are still a couple issues:

- you can notice there is a tab bar
- some sort of status bubble at the bottom
- language list is not populated (we have seen this in regular kiosk mode as well, so this is unrelated)
- the page is not clickable (possibly stuck in loading state ? as the status bubble seems to imply something is still being loaded from the server)

This was using the rather early post Python 3.12 live image: https://fedorapeople.org/groups/anaconda/webui_preview_image/x86_64/fedora-preview-live-installer-2023-07-11.iso

Thinking this might be some image related breakage, I've also tried using a regular Fedora Rawhide live image from July 12. This time the page content does not load at all (but its loading the correct page):

live_image_custom_theme_2
live_image_custom_theme_3
live_image_custom_theme_4

**HOW TO TEST**
The best option is to fetch a Live image - either one from the Image Builder pipeline or even a regular one.

Then you can add the following COPR which has packages corresponding to this PR as of right now:
https://copr.fedorainfracloud.org/coprs/m4rtink/anaconda-webui-firefox-theme/

Once you add the COPR, something like this should replace the on-media Anaconda with the one corresponding to this PR:

`sudo dnf install anaconda-999999999 --allowerasing`

Alternatively you can fetch the RPMs from the COPR (or build them yourself from the PR), `scp` them to the machine running your Live image & install them via DNF:

`sudo dnf install *.rpm --allowerasing`

Then start Anaconda in the usual way - either via its icon in the quick launch menu or via the `liveinst` script in `PATH`.

**Comment by VladimirSlavik** [comment]
Why does this contain two themes?

**Review by KKoukiou — COMMENTED** [review]
The default user and live user js code are mostly shared. Can't you create some common file and import it?

**Comment by M4rtinK** [comment]
> The default user and live user js code are mostly shared. Can't you create some common file and import it?

They are identical right now, but the expectation is the themes will diverge as we adjust the `default` (essentially boot.iso) theme, which for example does not need a header bar.

As for sharing code with the theme - I've just copied the theme created by @garrett in https://github.com/garrett/firefox-theme-frame & don't really have any experience with how it is done. So someone with a deeper knowledge of Firefox theming will need to answer this. :)

**Comment by M4rtinK** [comment]
> Why does this contain two themes?
The `live` theme is to be used for the Live images and the `default` theme is meant for everything else (but mainly boot.iso). Right now the themes are the same for both, as we don't have a boot.iso specific theme yet. In the future the `default` theme should be adjusted accordingly.

**Comment by M4rtinK** [comment]
Updated the PR to move most of the Firefox theme handling logic to the `webui-desktop` script, which seems to be a more suitable place for it.

**Comment by M4rtinK** [comment]
Unfortunately still stuck at the Web UI not (fully ?) loading:

webui_live_still_not_working

**Comment by garrett** [comment]
It looks like there are two title bars. First: window manager provided titlebar. Second: CSD with tabs. The theme was made with CSD (client-side decorations) in mind, not the window manager titlebar.

We have two options:

1. Turn off tab bar entirely and rely on the window manager. However, we could not disable, remove, or hide the close button.
2. Force CSD mode to remove the window manager titlebar. (But I thought the theme was doing this already? I need to double-check the `user.js` file, I guess?)

> This time the page content does not load at all (but its loading the correct page):

That looks like it's due to a problem in the JS in Anaconda web UI, not in the theme.

---

I've updated the theme to remove the button, fix the titlebar, and to try harder to make new windows show up.

But after playing with it more, I think a more proper approach would be to hide the tab bar and use the window manager titlebar... but we'd have a close button (unless we use the semi-restricted environment that GNOME Initial Setup does... which _may_ disable the X icon).

**Comment by KKoukiou** [comment]
The crash happens because the buildstamp file is not there. 
liveimg-crash


Is this file supposed to be always there? Should we handle it differently?

**Comment by garrett** [comment]
:ballot_box_with_check: I relicensed the theme to MIT, as asked.

**Comment by VladimirSlavik** [comment]
After thinking a bit about the themes, I realized the problem I have is that one is called "default". Perhaps "bootiso" would be better?

**Comment by M4rtinK** [comment]
> After thinking a bit about the themes, I realized the problem I have is that one is called "default". Perhaps "bootiso" would be better?

Well, the way it is we kinda need to always have a theme to pick - the live image theme is specifically picked for the Live image & the "default" theme will be used for everything else (which indeed is mostly the boot iso).

**Comment by M4rtinK** [comment]
> **TL;DR**: I've updated the theme. It should work much better now. (I hope!) wink

Thanks! Just updated the PR your the latest version of your theme - will post screenshots in a bit (& will generate a new boot.iso).

**Comment by M4rtinK** [comment]
live_image_new_theme

**Comment by M4rtinK** [comment]
Looking good IMHO. :)

**Comment by garrett** [comment]
Whew. I'm happy to see the theme working as intended @ https://github.com/rhinstaller/anaconda/pull/4918#issuecomment-1651902869

> It's a spaghetti monster, but it's flying!

:rofl:

File: ui/webui/webui-desktop [hunk 1]
```diff
@@ -61,7 +60,27 @@ case "$1" in
...
-BROWSER="/usr/bin/firefox --kiosk"
+# prepare empty firefox profile dir with theme based on the passed profile id
+FIREFOX_THEME_DIR="/usr/share/anaconda/firefox-theme"
+FIREFOX_PROFILE_PATH="/tmp/anaconda-firefox-profile"
+THEME_ID=$2
+
+# make sure the profile directory exists and is empty
+if [ -d ${FIREFOX_PROFILE_PATH} ]
+then
+    echo "Cleaning up existing Anaconda Firefox profile directory."
```
Comment (KKoukiou) [inline ui/webui/webui-desktop:71]: In what scenario would the profile directory on /tmp already exist?

**Review by KKoukiou — APPROVED** [review]
I have no comments on the css applied. Please attach to the PR the latest screenshot for both live and boot iso.

File: ui/webui/webui-desktop (hunk 1 above)
Comment (M4rtinK) [reply ui/webui/webui-desktop:71]: If you launch the installer multiple times on Live media - quite a frequent occurrence during debugging & we know from logs attached to bug reports, that users do it quite often in real life, usually when encountering an error and trying again.

In any case, we really should start with empty profile or else I would expect the behavior on followup starts being non-deterministic given the sheer amount of stuff Firefox dumps to the originally empty folder.
